_includes/tutorials/aggregating-minmax/flinksql/code/src/test/java/io/confluent/developer/AbstractFlinkKafkaTest.java

```diff
@@ -0,0 +1,171 @@
+   * @param schemaRegistryPort  the port that Schema Registry exposes
+   * @return resource file contents with port values substituted for placeholders
+   * @throws IOException if resource file can't be read
+   */
+  protected static String getResourceFileContents(
+      String resourceFileName,
+      Optional<Integer> kafkaPort,
+      Optional<Integer> schemaRegistryPort
+  ) throws IOException {
+    URL url = Resources.getResource(resourceFileName);
+    String contents = Resources.toString(url, StandardCharsets.UTF_8);
+    if (kafkaPort.isPresent()) {
+      contents = contents.replaceAll("KAFKA_PORT", kafkaPort.get().toString());
+    }
+    if (schemaRegistryPort.isPresent()) {
+      contents = contents.replaceAll("SCHEMA_REGISTRY_PORT", schemaRegistryPort.get().toString());
+    }
+    return contents;
+  }
+
+  /**
+   * Given a resource filename, return the resource file contents as a String.
+   *
+   * @param resourceFileName    the resource file name
+   * @return resource file contents
+   * @throws IOException if resource file can't be read
+   */
+  protected static String getResourceFileContents(
+      String resourceFileName
+  ) throws IOException {
+    // no Kafka / Schema Registry ports
+    return getResourceFileContents(resourceFileName, Optional.empty(), Optional.empty());
+  }
+
+  /**
+   * Utility method to convert a String containing multiple lines into a set of String's where
+   * each String is one line. This is useful for creating Flink SQL integration tests based on
+   * the tableau results printed via the Table API where the order of results is nondeterministic.
+   *
+   * @param s multiline String
+   * @return set of String's where each member is one line
+   */
+  protected static Set<String> stringToLineSet(String s) {
+    return Sets.newHashSet(Arrays.asList(s.split("\\r?\\n")));
+  }
+
+  /**
+   * Given a Flink Table API `TableResult` respresenting a SELECT statement result,
+   * capture and return the statement's tableau results.
+   *
+   * @param tableResult Flink Table API `TableResult` respresenting a SELECT statement result
+   * @return the SELECT statement's tableau results
+   */
+  protected static String tableauResults(TableResult tableResult) {
+    // The given table result may come from a table backed by the Kafka or Upsert Kafka connector,
+    // both of which perform unbounded (neverending) scans. So, in order to prevent tests from blocking
+    // on called to this method, we kick off a thread to kill the underlying job in 5 seconds.
+    //
+    // Note: as of Flink 1.17.0, the Kafka connector will support bounded scanning, which would obviate
+    // the need to do this. However, the Upsert Kafka connector will still be unbounded.
+    new Thread(() -> {
+      try {
+        Thread.sleep(5000);
+      } catch (InterruptedException e) {
+        // do nothing; proceed with job cancellation
+      }
+      tableResult.getJobClient().get().cancel();
+    }).start();
```

Comment:
I remember the conversation about the need to do this. One thought would be to put the ref to the `ByteArrayOutputStream` in the thread and run a loop checking for results as a way to make the test a little more deterministic and eliminate flaky failures.  But this works as well.

Reply:
good idea, changed to do this rather than sleep